code/modules/tgui/states.dm

```diff
@@ -0,0 +1,117 @@
+/**
+ * tgui states
+ *
+ * Base state and helpers for states. Just does some sanity checks, implement a state for in-depth checks.
+ */
+
+/**
+ * public
+ *
+ * Checks the UI state for a mob.
+ *
+ * required user mob The mob who opened/is using the UI.
+ * required state datum/ui_state The state to check.
+ *
+ * return UI_state The state of the UI.
+ */
+/datum/proc/tgui_status(mob/user, datum/tgui_state/state)
+	var/src_object = tgui_host(user)
+	. = STATUS_CLOSE
+	if(!state)
+		return
+
+	if(isobserver(user))
+		// If they turn on ghost AI control, admins can always interact.
+		if(user.client.advanced_admin_interaction)
+			. = max(., STATUS_INTERACTIVE)
+
+		// Regular ghosts can always at least view if in range.
+		var/clientviewlist = getviewsize(user.client.view)
+		if(get_dist(src_object, user) < max(clientviewlist[1],clientviewlist[2]))
+			. = max(., STATUS_UPDATE)
+
+	// Check if the state allows interaction
+	var/result = state.can_use_topic(src_object, user)
+	. = max(., result)
+
+/**
+ * private
+ *
+ * Checks if a user can use src_object's UI, and returns the state.
+ * Can call a mob proc, which allows overrides for each mob.
+ *
+ * required src_object datum The object/datum which owns the UI.
+ * required user mob The mob who opened/is using the UI.
+ *
+ * return UI_state The state of the UI.
+ */
+/datum/tgui_state/proc/can_use_topic(src_object, mob/user)
+	return STATUS_CLOSE // Don't allow interaction by default.
+
+/**
+ * public
+ *
+ * Standard interaction/sanity checks. Different mob types may have overrides.
+ *
+ * return UI_state The state of the UI.
+ */
+/mob/proc/shared_tgui_interaction(src_object)
+	if(!client) // Close UIs if mindless.
+		return STATUS_CLOSE
+	else if(stat) // Disable UIs if unconcious.
+		return STATUS_DISABLED
+	else if(incapacitated()) // Update UIs if incapicitated but concious.
+		return STATUS_UPDATE
+	return STATUS_INTERACTIVE
+
```

Comment:
wonder if there's an alternative way to do:

```
/mob/living/shared_ui_interaction(src_object)
	. = ..()
	if(!(mobility_flags & MOBILITY_UI) && . == UI_INTERACTIVE)
		return UI_UPDATE
```

May want to check the ocde history to see what this was before they added mobility flags.

Reply:
Will take a look

Reply:
Ignore this.